### DISCUSSION TITLE
How to join with a composite key?

### DISCUSSION THREAD

**Discussion by sandbox-apps** [supabase/supabase#2469] [Questions] [answered]
In examples we can join through
```js
const { data, error } = await supabase
  .from('products')
  .select(`
    id,
    supplier:supplier_id ( name ),
    purchaser:purchaser_id ( name )
  `)
```

But sometimes we also join using composite keys

**Comment by steve-chavez** [comment, marked as answer]
The foreign key constraint name can be used in that case

```js
const { data, error } = await supabase
  .from('products')
  .select(`
    id,
    supplier:products_supplier_id_fkey( name ),
    purchaser:products_purchaser_id_fkey( name )
  `)
```

It's also possible to rename the FK constraints, so the expression can be made shorter and save doing the alias(`supplier:`).

```sql
 alter table products
   rename constraint products_supplier_id_fkey  to supplier,
   rename constraint products_purchaser_id_fkey to purchaser;
```

```js
const { data, error } = await supabase
  .from('products')
  .select(`
    id,
    supplier( name ),
    purchaser( name )
  `)
```

(FK constraint names are local to the table - no collision, safe to rename them)

Note that specifying the column or the FK constraint is only needed in case where there's more than one FK relationship to the same table. With a single relationship, joins on composite keys should happen transparently.

For more details, see [postgrest embedding disambiguation](https://postgrest.org/en/stable/api.html#embedding-disambiguation).

**Reply by sandbox-apps** [reply]
Just to clarify..
Left hand side was the alias
Right hand side was the foreign key, and not the table name. (Or Im I still wrong)?

Here is another example from supabase
```js
const { data, error } = await supabase
  .from('countries')
  .select(`
    name,
    cities (
      name
    )
  `)
```

**Reply by sandbox-apps** [reply]
Just to cure my fixation
1. Left hand side was the alias
2. Right hand side was the foreign key or also table name
3. Foreign Key is a must if there is more than one foreign key on the table.
4. If there is only one, we can use the table name and it will automatically know how to join

**Reply by sandbox-apps** [reply]
@steve-chavez is there also a way to see all foreign key name at once in a table in supabase?

**Reply by steve-chavez** [reply]
> Left hand side was the alias
> Right hand side was the foreign key, and not the table name.. (Or Im I still wrong)?

Yes, correct.

> Foreign Key is a must if there is more than one foreign key on the table.

Only if your FKs are composite, otherwise you can use the column name(`supplier_id/purchaser_id` as per your example) to disambiguate the join.

> If there is only one, we can use the table name and it will automatically know how to join

Correct.

> is there also a way to see all foreign key name at once in a table in supabase?

Not through the UI, but you should be able to get them through the SQL Editor(with [this query](https://stackoverflow.com/a/1152321/4692662)).

Note that if there's a need to use the FK or a column name, there will be an error mentioning all the options to do the joins - see an example [here](https://github.com/supabase/supabase/discussions/2439#discussioncomment-1025620).

**Reply by sandbox-apps** [reply]
Really got me confused about when to use the table name vs column name at first. Thanks for clarifying.

**Reply by sandbox-apps** [reply]
So we can use the relationship value to disambiguate the query table

```js
{
      relationship: 'organisations_social_networks_fkey[social_networks][social_id]',
      origin: 'public.organisations',
      cardinality: 'm2o',
      target: 'public.social_networks'
    },
```
